### PR TITLE
vibrator: ff: add new Force Feedback vibrator

### PR DESCRIPTION
**Pull request by kcxt**
This vibrator type will search for an input device that supports the
FF_RUMBLE effect and use it.

Tested on the OnePlus 6T running Manjaro ARM.

### CODE REVIEW

**Comment by kcxt** [comment]
I haven't yet tested this on devices that don't have a force feedback device, I'll test on Ubuntu Touch / Halium and make sure it still works there. It would probably be good to have other people test too.

It might be best to try sysfs before force feedback, as sysfs is a lot more likely to succeed and I can just imagine downstream devices getting unhappy at having all their `input` devices queried...

**Review by mariogrip — APPROVED** [review]
LGTM :+1: will merge after ota16 is out

**Comment by kcxt** [comment]
@mariogrip the feedbackd implementation of force feedback devices requires a udev rule to "tag" the haptics device for feedbackd to use, for example the following is the line for the haptics in the OnePlus 6:
```
SUBSYSTEM=="input", KERNEL=="event*", ENV{ID_INPUT}=="1", SUBSYSTEMS=="input", ATTRS{name}=="pmi8998_haptics", TAG+="uaccess", ENV{FEEDBACKD_TYPE}="vibra"
```
This solution does require the device ship an extra udev rule, but it means that as the haptics device node was decided explicitly, rather than via guesswork it's much less likely to fail on some edge case, we force the device developer to declare the haptics device instead of potentially requiring patches to this service to handle such cases.

Whilst I can't necessarily think of a situation where this would fail, it does seem less robust. Do you think it would be worth reworking the detection to require a udev rule (and add relevant docs) or is this good enough?

As an aside, to the best of my knowledge the PinePhone haptics use FF, do haptics currently not work on ubports on the PinePhone? If not, can someone test this patch on the PinePhone before merging this?

**Comment by mariogrip** [comment]
> @mariogrip the feedbackd implementation of force feedback devices requires a udev rule to "tag" the haptics device for feedbackd to use, for example the following is the line for the haptics in the OnePlus 6:
> 
> ```
> SUBSYSTEM=="input", KERNEL=="event*", ENV{ID_INPUT}=="1", SUBSYSTEMS=="input", ATTRS{name}=="pmi8998_haptics", TAG+="uaccess", ENV{FEEDBACKD_TYPE}="vibra"
> ```
> 
> This solution does require the device ship an extra udev rule, but it means that as the haptics device node was decided explicitly, rather than via guesswork it's much less likely to fail on some edge case, we force the device developer to declare the haptics device instead of potentially requiring patches to this service to handle such cases.
> 
> Whilst I can't necessarily think of a situation where this would fail, it does seem less robust. Do you think it would be worth reworking the detection to require a udev rule (and add relevant docs) or is this good enough?
> 
> As an aside, to the best of my knowledge the PinePhone haptics use FF, do haptics currently not work on ubports on the PinePhone? If not, can someone test this patch on the PinePhone before merging this?

I think this is fine for now, the idea behind hfd is to be able to detect it automatically, manual adding is bad design imo. but we should add a override property if someone hit this edge case.

**Comment by Keneda-ut** [comment]
Since i switched to RC OTA-17, haptic feedback when (for example) entering pin code is quite unnoticeable.
It was good on OTA-16 and before.

Meizu MX4.

**Comment by kcxt** [comment]
> Since i switched to RC OTA-17, haptic feedback when (for example) entering pin code is quite unnoticeable.
> It was good on OTA-16 and before.
> 
> Meizu MX4.

@Keneda-ut there's plenty of reasons why that may happen, your best bet would be to open a new issue here.

**Comment by Keneda-ut** [comment]
@calebccff OK, anyway it seems to be back to normal since then, maybe after some reboots i don't know...

**Pull request merged**
